File: 2025_Oct/Egloo/2025-10-20-group-anagrams.py
class Solution(object):
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """

        dic = {}
        # key: hashValue , value : list

        for s in strs:
            hashValue = self.getHashValue(s)
            if not hashValue in dic:
                dic[hashValue] = []
            dic[hashValue].append(s)

        result = []
        for key in dic:
            result.append(dic[key])
        return result

    def getHashValue(self, word):
        return hash(''.join(sorted(word)))

File: 2025_Oct/youngwon/2025-10-20-Group-Anagrams.cpp
class Solution {
public:
      vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<string> sorted_strs = strs;  
        unordered_map<string, vector<int>> anagram_groups;

       
        for (int i = 0; i < sorted_strs.size(); ++i) {
            string sorted = sorted_strs[i];
            sort(sorted.begin(), sorted.end());
            anagram_groups[sorted].push_back(i);  
        }

        vector<vector<string>> result;
        for (const auto& [key, indices] : anagram_groups) {
            vector<string> group;
            for (int idx : indices) {
                group.push_back(strs[idx]);
            }
            result.push_back(move(group));
        }

        return result;